qiskit/utils/deprecation.py

```diff
@@ -50,30 +51,74 @@ def wrapper(*args, **kwargs):
+    kwargs: Dict[str, Any],
+    kwarg_map: Dict[str, str],
+    category: Type[Warning] = DeprecationWarning,
+) -> None:
+    func_name = func.__name__
     for old_arg, new_arg in kwarg_map.items():
+        if new_arg is None:
+            msg = (
+                f"{func_name} keyword argument {old_arg} is deprecated and "
+                "will in the future be removed."
+            )
+        else:
+            msg = (
+                f"{func_name} keyword argument {old_arg} is deprecated and "
+                f"replaced with {new_arg}."
+            )
+        _DeprecationMetadata.set_args_deprecation(func, arg=old_arg, msg=msg, since="TODO")
         if old_arg in kwargs:
             if new_arg in kwargs:
                 raise TypeError(f"{func_name} received both {new_arg} and {old_arg} (deprecated).")
+            warnings.warn(msg, category=category, stacklevel=3)
+            if new_arg is not None:
+                kwargs[new_arg] = kwargs.pop(old_arg)
 
-            if new_arg is None:
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    "will in future be removed.",
-                    category=category,
-                    stacklevel=3,
-                )
-            else:
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    f"replaced with {new_arg}.",
-                    category=category,
-                    stacklevel=3,
-                )
 
-                kwargs[new_arg] = kwargs.pop(old_arg)
+@dataclass(frozen=True)
+class _DeprecationMetadataEntry:
+    msg: str
+    since: str
+
+
+@dataclass
+class _DeprecationMetadata:
+    """Used to store deprecation information on a function.
+
+    This is used by the Qiskit Sphinx Theme to render deprecations in documentation. Warning:
+    coordinate changes with the Sphinx Theme's extension.
+    """
+
+    func_deprecation: Optional[_DeprecationMetadataEntry]
+    args_deprecations: Dict[str, _DeprecationMetadataEntry]
+
+    dunder_name: ClassVar[str] = "__qiskit_deprecation__"
+
+    @classmethod
+    def set_func_deprecation(cls, func: Callable, *, msg: str, since: str) -> None:
+        entry = _DeprecationMetadataEntry(msg, since)
+        if hasattr(func, cls.dunder_name):
+            metadata = cast(_DeprecationMetadata, getattr(func, cls.dunder_name))
+            metadata.func_deprecation = entry
+        else:
+            metadata = cls(func_deprecation=entry, args_deprecations={})
+            setattr(func, cls.dunder_name, metadata)
+
+    @classmethod
+    def set_args_deprecation(cls, func: Callable, *, arg: str, msg: str, since: str) -> None:
+        entry = _DeprecationMetadataEntry(msg, since)
+        if hasattr(func, cls.dunder_name):
+            metadata = cast(_DeprecationMetadata, getattr(func, cls.dunder_name))
+            metadata.args_deprecations[arg] = entry
```

Comment:
This is probably unnecessary complication, but do we care about what should happen if `func_deprecation` already exists in the dunder on `set_func_deprecation` (etc for a deprecated arg)?  I guess it would be indicative of a Qiskit dev doing something wrong, such as accidentally wrapping a function more than once.

Reply:
This will handle that correctly :) Here, we're only modifying the `args_deprecations` field and not touching `func_deprecation`, regardless of it's `None` or a `_DeprecationMetadataEntry`.

I'm planning on writing tests today to ensure this logic is all right. Because I actually have found a few examples of where we both deprecate the function and its arguments at the same time.

Reply:
Sorry, I meant what if someone tries to do the function one twice, or the same argument twice. The current behaviour is "silently override".

Reply:
Oh, yeah. That would indeed be Operator Error and I don't think we need to worry about it. 

Afaict, the only way to do this is literally put the decorator twice on the same function. Code review should catch that.